lib/nx_inflate.c

```diff
@@ -2046,58 +2052,160 @@ int inflateInit2_(z_streamp strm, int windowBits, const char *version, int strea
+			rc = hw_reset(strm);
+			if (rc != Z_OK)
+				return rc;
+			sme->hw_state->switchable = 1;
+		} else {
+			/* hw_state is NULL because inflateInit was not able to
+			   allocate it, most likely due to lack of credits on
+			   PowerVM. Let's try again. */
+			   strm->state = NULL;
+			   rc = nx_inflateInit2_(strm, sme->windowBits,
+						 sme->version, sme->stream_size);
+			   if (rc != Z_OK) {
+				   /* Just keep using zlib */
+				   strm->state = sme->sw_state;
+				   return Z_OK;
+			   }
+
+			   nx_streamp s = (nx_streamp) strm->state;
+			   sme->hw_state = s;
+			   s->sw_stream = sme->sw_state;
+			   s->switchable = 1;
+		}
+		/* Leave NX state 'loaded' on strm */
+	} else if (!has_nx_state(strm)) {
+		rc = sw_reset(strm);
+	} else {
+		rc = hw_reset(strm);
 	}
 
 	return rc;
 }
 
+int inflateReset(z_streamp strm)
+{
+	return inflateReset_skeleton(strm, sw_inflateReset, nx_inflateReset);
+}
+
 int inflateResetKeep(z_streamp strm)
+{
+	return inflateReset_skeleton(strm, sw_inflateResetKeep,
+				     nx_inflateResetKeep);
+}
+
+int inflateReset2(z_streamp strm, int windowBits)
 {
 	int rc;
 
-	if (0 == has_nx_state(strm)){
-		rc = sw_inflateResetKeep(strm);
-	}else{
-		rc = nx_inflateResetKeep(strm);
+	if (nx_config.mode.deflate == GZIP_AUTO) {
+		struct stream_map_entry *sme;
+
+		if (nx_map_get(stream_map, strm, (void **)&sme)) {
+			prt_err("Trying to reset uninitialized stream\n");
+			return Z_STREAM_ERROR;
+		}
+		validate_stream_map_entry(sme, strm);
+
+		strm->state = sme->sw_state;
+		rc = sw_inflateReset2(strm, windowBits);
+		if (rc != Z_OK)
+			return rc;
+
+		if (sme->hw_state != NULL) {
+			strm->state = (void *) sme->hw_state;
+			rc = nx_inflateReset2(strm, windowBits);
+			if (rc != Z_OK)
+				return rc;
+			sme->hw_state->switchable = 1;
+		} else {
+			/* hw_state is NULL because inflateInit was not able to
+			   allocate it, most likely due to lack of credits on
+			   PowerVM. Let's try again. */
+			   strm->state = NULL;
+			   rc = nx_inflateInit2_(strm, sme->windowBits,
+						 sme->version, sme->stream_size);
+			   if (rc != Z_OK) {
+				   /* Just keep using zlib */
+				   strm->state = sme->sw_state;
+				   return Z_OK;
+			   }
+
+			   nx_streamp s = (nx_streamp) strm->state;
+			   sme->hw_state = s;
+			   s->sw_stream = sme->sw_state;
+			   s->switchable = 1;
+		}
+		/* Leave NX state 'loaded' on strm */
+	} else if (!has_nx_state(strm)) {
+		rc = sw_inflateReset2(strm, windowBits);
+	} else {
+		rc = nx_inflateReset2(strm, windowBits);
 	}
 
 	return rc;
 }
 
-
 int inflateEnd(z_streamp strm)
 {
 	int rc;
 
 	/* statistic */
 	zlib_stats_inc(&zlib_stats.inflateEnd);
 
-	if (0 == has_nx_state(strm)){
+	if (nx_config.mode.deflate == GZIP_AUTO) {
+		struct stream_map_entry *sme = nx_map_remove(stream_map, strm);
+		validate_stream_map_entry(sme, strm);
+
+		/* TODO: What to do about the delay? */
+		/* if (0 != avg_delay) */
+		/* 	decrease_delay(); */
```

Comment:
Why not decrease it when there is no nx state?

Reply:
Agreed. I did this on nx_deflate.c but forgot to do the same here. Thanks!